File: 250320/겹치지 않게 선분 고르기/select-segments-without-overlap.cpp
#include <bits/stdc++.h>
using namespace std;

int n, ans;
vector<pair<int, int>> line;
vector<pair<int, int>> cho;

bool Check() {

    for(int i=1; i<(int)cho.size(); i++) {
        int x, y;
        tie(x, y) = cho[i-1];

        int x2, y2;
        tie(x2, y2) = cho[i];

        if(!(y < x2)) return false;
    }

    return true;
}

void FindMaxLine(int idx) {
    if(idx == (int)line.size()) {
        if(Check()) {
            ans = max(ans, (int)cho.size());
        }
        return;
    }

    cho.push_back(line[idx]);
    FindMaxLine(idx + 1);
    cho.pop_back();

    FindMaxLine(idx+1);
}

int main() {
    cin >> n;

    for(int i=0; i<n; i++) {
        int x, y;
        cin >> x>>y;
        line.push_back({x, y});
    }

    sort(line.begin(), line.end());
    FindMaxLine(0);

    cout << ans;
    return 0;
}

File: 250325/겹치지 않게 선분 고르기/select-segments-without-overlap.py
n = int(input())

lines = [
    tuple(map(int, input().split()))
    for _ in range(n)
]

lines.sort()
selc_line = []
ans = 0

def check():
    for i in range(1, len(selc_line)):
        x1, y1 = selc_line[i-1]
        x2, y2 = selc_line[i]

        if x2 <= y1: 
            return False
    
    return True

def findMaxLine(idx):
    if idx == len(lines):
        if check():
            global ans

            ans = max(ans, len(selc_line))
        return
    
    selc_line.append(lines[idx])
    findMaxLine(idx+1)
    selc_line.pop()

    findMaxLine(idx+1)


findMaxLine(0)

print(ans)